docs/architecture/adr-110-remote-mempool.md

```diff
@@ -0,0 +1,364 @@
+# this section will be ignored.
+#
+# A remote mempool is only usable by a validator node. Turning on the remote
+# mempool for a full node will simply disable any mempool-related functionality
+# on that full node, and the full node will not interact with any mempool at
+# all.
+#
+[mempool.remote]
+# The base URL for the gRPC interface to the remote mempool.
+url = "http://localhost:28880"
+
+# The timeout for fetching transaction data during PrepareProposal. Exceeding
+# this timeout will result in a panic.
+timeout_fetch = "5s"
+
+# The timeout for reaping (removing) transaction data after a block has been
+# committed. Exceeding this timeout will result in a panic.
+timeout_reap = "1s"
+
+# The timeout for initiating the TxsAvailable call.
+timeout_txs_available = "1s"
+
+# The maximum number of times to retry calling the TxsAvailable streaming
+# endpoint if it fails. Exceeding this number of retries (e.g. if the connection
+# to the remote mempool is down) will result in a panic.
+txs_available_retries = 3
+```
+
+### RPC API
+
+At the time of this writing, it is recommended to implement a gRPC-based RPC
+standard for interacting with the remote mempool (which uses Protobuf encoding
+over HTTP/2) because:
+
+- gRPC, as a standard, is well-established and code generation technologies are
+  widely available for many different programming languages, allowing mempool
+  developers more flexibility in how they build those mempools.
+- Comet is moving toward making more extensive use of gRPC in future.
+- Load balancing technologies for HTTP/2 using reverse proxies such as nginx are
+  relatively well-established, making routing of requests to scaled-out mempool
+  clusters relatively straightforward.
+
+An alternative here would be to implement a RESTful HTTP/3 API (using QUIC), but
+HTTP/3 support in reverse proxies is still in its early stages. Such a protocol
+can be considered in a future iteration of the interface if it provides
+substantial performance/latency benefits.
+
+The following gRPC API is proposed.
+
+```protobuf
+syntax = "proto3";
+package tendermint.services.remote_mempool.v1;
+
+// RemoteMempoolService is implemented by a remote mempool instance. A CometBFT
+// validator makes use of a client generated from this interface definition to
+// interact with the remote mempool.
+//
+// It is up to the implementor of the remote mempool to define transaction
+// submission mechanisms/interfaces.
+service RemoteMempoolService {
+    // Fetch allows a CometBFT validator to obtain the next batch of
+    // transactions to be included during PrepareProposal when the current node
+    // is a proposer.
+    rpc Fetch(FetchRequest) returns (FetchResponse);
+
+    // Reap takes a set of transaction keys and removes the corresponding
+    // transactions from the mempool.
+    //
+    // This will only be called after the block is committed by consensus.
+    rpc Reap(ReapRequest) returns (ReapResponse);
```

Comment:
The name `Reap` suggest that the mempool should return something, but `ReapResponse` is always empty. Maybe rename it to something like `RemoveTxs` or `GarbageCollectTxs`?